feat(helm-prereqs): automate DPF (DOCA Platform Framework) install in setup.sh (#3578)
NVIDIA/bare-metal-manager-core: docs/manuals/dpf.md

---

> ## Automated install (default) — `setup.sh`

Reviewer: Generally best to avoid formatting and most punctuation in headings
```suggestion
## Automated install (default) with setup.sh
```

For clusters bootstrapped with `helm-prereqs/setup.sh`, DPF installs **by
default** — the entire DPF installation **and** carbide-api enablement below is
automated end-to-end. The DPF operator stack installs as **phase 5b** (before
NICo Core); carbide-api DPF enablement happens as **phase 6b** (after Core).
Pass `--skip-dpf` (or `NICO_SKIP_DPF=true`) to opt out — e.g. sites with no
DPUs, or that still use the deprecated iPXE DPU path.

```bash
export NICO_DPF_DPU_INTERFACE=<controller-interface>   # keepalived interface for the DPU cluster VIP
export NICO_DPF_DPU_CLUSTER_VIP=<vip>                  # VIP the DPUs use to reach their control plane
export NICO_DPF_BMC_ROOT_PASSWORD=<bmc-password>       # site-wide BMC root (see "BMC root precondition")
export NICO_DPF_METALLB_POOL=<pool>                    # optional: MetalLB pool advertising the VIP
./setup.sh                                             # DPF installs by default; add --skip-dpf to opt out
```

What the run does, mapped to the manual sections below:

| Manual section | Automated by setup.sh |

> | --- | --- |

Reviewer: ```suggestion
The following table maps the sections on this page to what the run does:

| Manual section | Automated by setup.sh |
| -------------- | --------------------- |
```

| §3.1 RBAC | Created by the NICo Core chart (`nico-api.dpf.rbacCreate=true`, set automatically) — the Role/RoleBinding subject is the chart's actual ServiceAccount. |
| §3.2–3.4 CRs | `DPFOperatorConfig` (API VIP/port derived from the `kubernetes` Endpoints unless `NICO_DPF_K8S_API_VIP/PORT` are set), `DPUCluster`, and the optional VIP LoadBalancer Service are applied from `helm-prereqs/operators/dpf/`. |
| §3.5 site config + §4 enablement | **Two-phase (phase 6b).** carbide-api's DPF SDK requires the site-wide BMC root password at startup, which can only be set through a running carbide-api — so setup.sh deploys Core with `[dpf]` **off**, sets the BMC root password via `nico-admin-cli` (see below), upgrades Core to `[dpf]` **on**, then **restarts carbide-api** (the upgrade only rewrites the ConfigMap; `[dpf]` is read at startup only) so its DPF SDK initializes and creates the BFB, DPUFlavor, and DPUDeployment. |

Per-host enablement (§3.6) and the CLI appendix still apply unchanged. The
sections below remain the reference for what is being installed, for manual

> installs, and for environments not using `setup.sh`.

Reviewer: Adding links to each section, as well as some breaks for easier reading in the rendered table.
```suggestion
| §1.1 [Namespace](#11-create-the-dpf-operator-namespace) | Creates `dpf-operator-system`. |
| §1.2 [Secrets](#12-image-pull-and-helm-repository-credentials) | Creates `hbn-user-password` (generated once), `dpf-pull-secret` / `nico-pull-secret` (from `NICO_DPF_NGC_API_KEY` / `NICO_DPF_NICO_NGC_API_KEY`, defaulting to `REGISTRY_PULL_SECRET`), and the three Argo CD repository Secrets (GA repo URLs; override with `NICO_DPF_HELM_REPO_OCI/HTTPS/CARBIDE`). |
| §1.4 [Cert-manager policy](#14-cert-manager-policy-and-rbac-for-dpf) | Applied only when the approver-policy CRD exists. The stock helm-prereqs cert-manager does **not** run approver-policy, so this is normally skipped. |
| [Prerequisites](#1-prerequisites) (Argo CD, Kamaji, NFD, maintenance-operator) | Installed from `helm-prereqs/helmfile.yaml` with versions/values pinned from `doca-platform deploy/helmfiles/prereqs.yaml`; cert-manager and local-path-provisioner are reused from the base install.<br/><br/>Kamaji has a cold-start deadlock (its controller needs the `default` DataStore, whose admission webhook the controller itself serves) — setup.sh breaks it automatically. Note: upstream pins cert-manager v1.19.3 while helm-prereqs ships v1.17.1 — a known, compatible skew. |
| §2 [Operator install](#2-dpf-installation) | Clones `NVIDIA/doca-platform` at `NICO_DPF_VERSION` (default `v26.4.0`, cached under `helm-prereqs/.dpf-src/`) and installs `deploy/charts/dpf-operator`.<br/><br/>The in-repo source chart ships an empty `controllerManager.image`, so setup.sh sets it to `nvcr.io/nvidia/doca/dpf-system:$NICO_DPF_VERSION` (override with `NICO_DPF_IMAGE_REPO`).<br/><br/>The GA `nvidia/doca` images are **public**, so they pull anonymously by default — a registry-scoped pull secret without `nvidia/doca` entitlement makes nvcr.io 403 the pull. Set `NICO_DPF_IMAGE_PULL_SECRET` only for private DPF/DOCA registries. |
| §3.1 [RBAC](#31-rbac-for-the-nico-orchestrator) | Created by the NICo Core chart (`nico-api.dpf.rbacCreate=true`, set automatically) — the Role/RoleBinding subject is the chart's actual ServiceAccount. |
| §3.2–3.4 CRs  | [DPFOperatorConfig](#32-dpfoperatorconfig) (API VIP/port derived from the `kubernetes` Endpoints unless `NICO_DPF_K8S_API_VIP/PORT` are set), [DPUCluster](#33-dpucluster), and the optional [VIP LoadBalancer Service](#34-vip-loadbalancer-service-and-endpoints) are applied from `helm-prereqs/operators/dpf/`. |
| §3.5 [Site config](#35-enable-dpf-in-the-nico-site-config) + §4 [Enablement](#4-restart-carbide-api-to-create-the-dpf-initialization-objects) | **Two-phase (phase 6b).** Carbide-api's DPF SDK requires the site-wide BMC root password at startup, which can only be set through a running carbide-api.<br/><br/>`setup.sh` deploys Core with `[dpf]` **off**, sets the BMC root password via `nico-admin-cli` (see below), upgrades Core to `[dpf]` **on**, then **restarts carbide-api**.<br/><br/>The upgrade only rewrites the ConfigMap; `[dpf]` is read at startup only. The restart ensures that the DPF SDK initializes and creates the BFB, DPUFlavor, and DPUDeployment. |

[Per-host enablement](#36-mark-hosts-as-dpf-managed-in-expected-machines) (§3.6) and the [CLI appendix](#appendix-nico-admin-cli-dpf-command-reference) still apply unchanged. The sections below remain the reference for what is being installed, for manual installs, and for environments not using `setup.sh`.
```

### BMC root precondition (why the enablement is two-phase)

carbide-api's DPF SDK init **hard-requires** the site-wide BMC root credential
(the shared BMC password DPF uses to reach the DPUs' host BMCs over Redfish). If
it is not set, carbide-api fails to start with

> `Failed to initialize DPF SDK: ... Site wide BMC root credentials not set`.

Author: Already hyphenated in the current text (all occurrences read "site-wide").

That credential can only be set through a **running** carbide-api (the
`SetBmcRootPassword` RPC / `nico-admin-cli credential add-bmc`), so DPF cannot be
enabled on the very first Core deploy — hence the two-phase flow.

setup.sh handles this automatically (DPF is the default): it issues a short-lived
admin client cert from the `nicoca` PKI (see
[ingesting-hosts.md](../provisioning/ingesting-hosts.md)) and runs
`nico-admin-cli credential add-bmc --kind=site-wide-root` from an in-cluster Job
(the CLI ships in the NICo image at `/opt/carbide/nico-admin-cli`), reaching
carbide-api through its external LoadBalancer. `NICO_DPF_BMC_ROOT_PASSWORD` is
therefore **required** unless `--skip-dpf`.

> **DPU-less clusters.** The DPF operator, Kamaji `DPUCluster`, and carbide-api
> all come up, but `DPFOperatorConfig` stays `Ready=False` until its DPU-side
> services (multus, flannel, sriov-device-plugin, ovs-cni, sfc-controller, …)
> schedule — which needs actual DPU nodes. On a cluster with no BlueField

> > hardware this is expected, not an error.

Reviewer: Using the Fern component here
```suggestion
<Info title="Clusters without DPUs">
The DPF operator, Kamaji `DPUCluster`, and carbide-api all come up, but `DPFOperatorConfig` stays `Ready=False` until its DPU-side services (multus, flannel, sriov-device-plugin, ovs-cni, sfc-controller, and so on) schedule — which needs actual DPU nodes. On a cluster with no BlueField hardware this is expected, not an error.
</Info>
```

DPF pulls several Helm charts via Argo CD. Apply the following Secrets so that
Argo CD can authenticate to the NGC Helm repositories:

> **Note**: the Secrets must live in the namespace where Argo CD is installed,
> and `overrides.argoCDNamespace` in the `DPFOperatorConfig` (§3.2) must match.
> The examples below use `argocd`; `setup.sh` (DPF default) installs Argo CD
> into `dpf-operator-system` (same as upstream's prereqs helmfile) and creates

> > these Secrets there.

Reviewer: Fern component
```suggestion
<Note>
Secrets must live in the namespace where Argo CD is installed, and `overrides.argoCDNamespace` in the [DPFOperatorConfig](#32-dpfoperatorconfig) must match. The examples below use `argocd`; `setup.sh` (DPF default) installs Argo CD into `dpf-operator-system` (same as upstream's prereqs helmfile) and creates these Secrets there.
</Note>
```

**before NICo is started**. They configure the DPF operator for NICo's
provisioning model and grant the orchestrator the access it needs.

### 3.1. RBAC for the NICo orchestrator

The NICo orchestrator's ServiceAccount needs access to the DPF custom
resources in `dpf-operator-system`. This is a **namespaced Role +
RoleBinding** scoped to `dpf-operator-system` — cluster-admin is *not*
required.

> **Note**: On chart-based deployments this Role/RoleBinding is created by the
> NICo Core chart itself (`nico-api.dpf.rbacCreate=true`, set automatically by
> `setup.sh`, DPF default), with the subject bound to the chart's actual
> ServiceAccount (`nico-api` in `nico-system` by default). Apply the manifest
> below only on deployments that do not use the chart — and adjust the
> RoleBinding subject to your deployment's ServiceAccount name and namespace

> > (e.g. `carbide-api` / `forge-system` on legacy carbide-named deployments).

Reviewer: ```suggestion
</Note>
On chart-based deployments, this Role/RoleBinding is created by the
NICo Core chart itself (`nico-api.dpf.rbacCreate=true`, set automatically by `setup.sh`, and the DPF default), with the subject bound to the chart's actual ServiceAccount (`nico-api` in `nico-system` by default).

Apply the manifest below only on deployments that do not use the chart, and adjust the RoleBinding subject to your deployment's ServiceAccount name and namespace (for example, `carbide-api` / `forge-system` on legacy carbide-named deployments).
</Note>
```

Whether a given host is provisioned via DPF or via iPXE is decided per host,
in the *expected machines* list that NICo loads on startup. The relevant
field is **`dpf_enabled`** on each expected-machine entry. A host is
provisioned via DPF only when **both** of the following are true:

1. `[dpf].enabled = true` in the site config (section 3.5), and
2. `dpf_enabled = true` on that host's expected-machine entry.

> **Per-host default is `true`.** When `dpf_enabled` is omitted, it is stored
> as `true` — DPF is the default provisioning path (iPXE is deprecated). Set

> > `dpf_enabled = false` explicitly to keep a host on iPXE.

Reviewer: ```suggestion
**The per-host default is `true`.** When `dpf_enabled` is omitted, it is stored as `true` — DPF is the default provisioning path (iPXE is deprecated). To keep a host on iPXE, set `dpf_enabled = false`.
```

> **Important**: this is **not a merge**. Any expected-machine row that is
> not present in the file is **deleted**. Each entry is then re-created via
> the same path as `add`, so any entry whose `dpf_enabled` is omitted is
> re-inserted with `dpf_enabled = true` (the default). Set `dpf_enabled = false`

> > explicitly to keep a host on iPXE.

Reviewer: ```suggestion
<Warning>
This is **not a merge**. Any expected-machine row that is not present in the file is **deleted**. Each entry is then re-created via the same path as `add`, so any entry whose `dpf_enabled` is omitted is re-inserted with `dpf_enabled = true` (the default). Set `dpf_enabled = false` explicitly to keep a host on iPXE.
</Warning>
```

Sets `machines.dpf.enabled = true` on the given host's runtime row by calling
the `ModifyDPFState` RPC.

### `dpf disable` — turn DPF off for a host

```bash
nico-admin-cli dpf disable <host-machine-id>
```

| Argument | Required | Notes |

> |---|:---:|---|

Reviewer: ```suggestion
| Argument | Required | Notes |
| -------- | :------: | ----- |
```

> | `<host-machine-id>` | yes | Must be a **host** machine id; DPU ids are rejected. |

Reviewer: ```suggestion
| `<host-machine-id>` | yes | Must be a **host** machine id; DPU IDs are rejected. |
```

| Goal | Command |
| --- | --- |
| Turn DPF on for an already-discovered host (transient) | `nico-admin-cli dpf enable <host-id>` |
| Turn DPF off for a host (refused after DPF ingestion) | `nico-admin-cli dpf disable <host-id>` |
| Show DPF state for one host | `nico-admin-cli dpf show <host-id>` |
| List DPF state for all hosts | `nico-admin-cli dpf show` |
| Snapshot DPF CRs for a host | `nico-admin-cli dpf snapshot <host-id>` |
| Diff configured vs. deployed DPF service versions | `nico-admin-cli dpf service-version` |

---

## Appendix: Building the NICo DPUService images and charts

The mandatory DPUServices that carbide-api deploys through DPF

> (`carbide-dpu-agent`, `carbide-dhcp-server`, `carbide-fmds`,

Reviewer: ```suggestion
(`carbide-dpu-agent`, `carbide-dhcp-server`, `carbide-fmds`, and
```

`carbide-otelcol`) are built from this repository under `bluefield/`. These are
NICo's own images and belong in the **same registry you push Core/REST to** —
build and push them there, then point `[dpf.services.*]` at them.

```bash
cd bluefield
export CARBIDE_IMAGE_REGISTRY=<your-registry>   # default nvcr.io/nvidia/carbide
export DPU_AGENT_PKG_VERSION=<tag>              # the image/chart version

# arm64 container images (forge-dpu-agent, forge-dhcp-server, carbide-fmds,
# forge-dpu-otel-agent, otelcol-contrib)
cargo make docker-build-all

# Helm chart packages for the matching DPUService charts
cargo make helm-package-all

# then docker push / helm push each to <your-registry>
```

> The chart sources live in `bluefield/charts/` (`nico-dpu-agent`,

Reviewer: ```suggestion
The chart sources are in `bluefield/charts/` (`nico-dpu-agent`,
```

> `nico-dhcp-server`, `nico-fmds`, `nico-otelcol`); each `values.yaml` starts

Reviewer: ```suggestion
`nico-dhcp-server`, `nico-fmds`, `nico-otelcol`); each values file starts
```

with the `### DPF contract ###` block that DPF/carbide-api populates at
deploy time. Point carbide-api at your pushed images/charts via the

> `[dpf.services.<svc>]` blocks (§3.5): `helm_repo_url`, `helm_chart`,

Reviewer: ```suggestion
`[dpf.services.<svc>]` blocks ([refer to section 3.5]((#35-enable-dpf-in-the-nico-site-config))): `helm_repo_url`, `helm_chart`,
```

`helm_version`, `docker_repo_url`, `docker_image_tag`, and
`docker_image_pull_secret = "nico-pull-secret"`. Unset services fall back to
the built-in defaults (public NGC for `dts`/`doca_hbn`; private `carbide-dev`,
needs `nico-pull-secret`, for the NICo DPUService images).

> The DTS (`doca-telemetry`) and `doca-hbn` services, and the DPF operator +

Reviewer: ```suggestion
The DTS (`doca-telemetry`) and `doca-hbn` services, and the DPF operator and
```

operand images, are NVIDIA-published on NGC and **pull anonymously by default**
— no build or registry needed. To mirror them into your own registry (air-gapped

> or one-registry setups), see

Reviewer: ```suggestion
or one-registry setups), refer to
```
